File: android/src/main/java/expo/modules/videokit/ExpoVideoKitModule.kt
package expo.modules.videokit

import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.core.os.bundleOf
import androidx.media3.common.util.UnstableApi
import dev.videokit.VideoKit
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import kotlin.math.max
import kotlin.math.min

@UnstableApi
class ExpoVideoKitModule : Module() {
  private val videoKit = VideoKit()
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun definition() = ModuleDefinition {
    Name("ExpoVideoKit")

    Events("onProgress")

    AsyncFunction("convert") Coroutine { inputUri: String, outputPath: String, config: ConversionConfigRecord ->
      val context = appContext.reactContext
        ?: throw CodedException("React context is unavailable")

      val resolvedOutputPath = resolveOutputPath(outputPath)
      val inputFileSize = resolveFileSize(context, inputUri)
      val progressReporter = ProgressReporter { progress ->
        sendEvent("onProgress", bundleOf("progress" to progress.toDouble()))
      }

      progressReporter.startPollingOutputFile(resolvedOutputPath, inputFileSize, mainHandler)

      val result = videoKit.convert(
        context = context,
        inputUri = Uri.parse(inputUri),
        outputPath = resolvedOutputPath,
        config = config.toNative(),
        onProgress = { progress ->
          progressReporter.updateNativeProgress(progress)
        },
      )

      result.getOrElse { throw CodedException(it.message ?: "Video conversion failed", it) }
      progressReporter.finish()
    }

    AsyncFunction("upload") Coroutine { filePath: String, url: String, config: UploadConfigRecord ->
      val resolvedFilePath = resolveOutputPath(filePath)
      val progressReporter = ProgressReporter { progress ->
        sendEvent("onProgress", bundleOf("progress" to progress.toDouble()))
      }

      val result = videoKit.upload(
        filePath = resolvedFilePath,
        url = url,
        config = config.toNative(),
        onProgress = { progress ->
          progressReporter.updateNativeProgress(progress)
        },
      )

      result.getOrElse { throw CodedException(it.message ?: "Video upload failed", it) }
      progressReporter.finish()
    }

    AsyncFunction("convertAndUpload") Coroutine {
      inputUri: String,
      uploadUrl: String,
      outputPath: String?,
      conversionConfig: ConversionConfigRecord,
      uploadConfig: UploadConfigRecord,
    ->
      val context = appContext.reactContext
        ?: throw CodedException("React context is unavailable")

      val resolvedOutputPath = outputPath?.let(::resolveOutputPath)
      val inputFileSize = resolveFileSize(context, inputUri)
      val progressReporter = ProgressReporter { progress ->
        sendEvent("onProgress", bundleOf("progress" to progress.toDouble()))
      }

      if (resolvedOutputPath != null) {
        progressReporter.startPollingOutputFile(resolvedOutputPath, inputFileSize, mainHandler)
      }

      val result = videoKit.convertAndUpload(
        context = context,
        inputUri = Uri.parse(inputUri),
        uploadUrl = uploadUrl,
        outputPath = resolvedOutputPath,
        conversionConfig = conversionConfig.toNative(),
        uploadConfig = uploadConfig.toNative(),
        onProgress = { progress ->
          progressReporter.updateNativeProgress(progress)
        },
      )

      result.getOrElse {
        throw CodedException(it.message ?: "Video convert and upload failed", it)
      }
      progressReporter.finish()
    }
  }
}

private class ProgressReporter(
  private val emit: (Float) -> Unit,
) {
  private var lastReported = -1f
  private var nativeProgress = 0f
  private var pollRunnable: Runnable? = null
  private var pollHandler: Handler? = null

  fun startPollingOutputFile(outputPath: String, inputFileSize: Long?, handler: Handler) {
    if (inputFileSize == null || inputFileSize <= 0L) {
      return
    }

    pollHandler = handler
    val runnable = object : Runnable {
      override fun run() {
        val outputSize = File(outputPath).length()
        if (outputSize > 0L) {
          val estimated = min(0.95f, outputSize.toFloat() / inputFileSize.toFloat() * 1.1f)
          report(max(nativeProgress, estimated))
        }
        pollHandler?.postDelayed(this, 250)
      }
    }

    pollRunnable = runnable
    handler.postDelayed(runnable, 250)
  }

  fun updateNativeProgress(progress: Float) {
    nativeProgress = progress
    report(progress)
  }

  fun finish() {
    pollRunnable?.let { runnable ->
      pollHandler?.removeCallbacks(runnable)
    }
    pollRunnable = null
    pollHandler = null
    report(1f)
  }

  private fun report(progress: Float) {
    val clamped = min(max(progress, 0f), 1f)
    if (clamped <= lastReported + 0.005f && clamped < 1f) {
      return
    }

    lastReported = clamped
    emit(clamped)
  }
}

private fun resolveFileSize(context: android.content.Context, inputUri: String): Long? {
  return try {
    context.contentResolver.openAssetFileDescriptor(Uri.parse(inputUri), "r")?.use { descriptor ->
      descriptor.length.takeIf { it > 0L }
    }
  } catch (_: Exception) {
    null
  }
}

File: ios/ExpoVideoKitModule.swift
import ExpoModulesCore
import VideoKit

struct ConversionConfigRecord: Record {
  @Field var targetWidth: Int = 1080
  @Field var targetHeight: Int = 1920
  @Field var targetFps: Int = 15
  @Field var videoMimeType: String = "video/avc"
  @Field var portraitEncodingEnabled: Bool = true

  func toNative() -> ConversionConfig {
    ConversionConfig(
      targetWidth: targetWidth,
      targetHeight: targetHeight,
      targetFps: targetFps,
      videoMimeType: videoMimeType,
      portraitEncodingEnabled: portraitEncodingEnabled
    )
  }
}

struct UploadConfigRecord: Record {
  @Field var contentType: String = "video/mp4"
  @Field var method: String = "PUT"

  func toNative() -> UploadConfig {
    UploadConfig(contentType: contentType, method: method)
  }
}

private func resolveURL(from uri: String) throws -> URL {
  let trimmed = uri.trimmingCharacters(in: .whitespacesAndNewlines)
  if trimmed.hasPrefix("file://") {
    guard let url = URL(string: trimmed) else {
      throw Exception(
        name: "InvalidInputUri",
        description: "Invalid input URI: \(uri)"
      )
    }
    return url
  }

  if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
    guard let url = URL(string: trimmed) else {
      throw Exception(
        name: "InvalidInputUri",
        description: "Invalid input URI: \(uri)"
      )
    }
    return url
  }

  return URL(fileURLWithPath: trimmed)
}

private func resolveOutputPath(_ path: String) -> String {
  let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
  if trimmed.hasPrefix("file://"), let url = URL(string: trimmed) {
    return url.path
  }
  return trimmed
}

private func describeError(_ error: Error) -> String {
  let localized = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
  if !localized.isEmpty {
    return localized
  }
  return String(describing: error)
}

private final class ProgressReporter {
  private let emit: (Double) -> Void
  private var lastReported: Double = -1

  init(emit: @escaping (Double) -> Void) {
    self.emit = emit
  }

  func updateNativeProgress(_ progress: Float) {
    report(Double(progress))
  }

  func finish() {
    report(1)
  }

  private func report(_ progress: Double) {
    let clamped = min(max(progress, 0), 1)
    guard clamped > lastReported + 0.005 || clamped >= 1 else {
      return
    }

    lastReported = clamped
    if Thread.isMainThread {
      emit(clamped)
    } else {
      DispatchQueue.main.async {
        self.emit(clamped)
      }
    }
  }
}

public class ExpoVideoKitModule: Module {
  private let videoKit = VideoKit()

  public func definition() -> ModuleDefinition {
    Name("ExpoVideoKit")

    Events("onProgress")

    AsyncFunction("convert") {
      (inputUri: String, outputPath: String, config: ConversionConfigRecord) async throws in
      let inputURL = try resolveURL(from: inputUri)
      let resolvedOutputPath = resolveOutputPath(outputPath)
      let progressReporter = ProgressReporter { progress in
        self.sendEvent("onProgress", ["progress": progress])
      }

      let result = await self.videoKit.convert(
        inputURL: inputURL,
        outputPath: resolvedOutputPath,
        config: config.toNative(),
        onProgress: { progress in
          progressReporter.updateNativeProgress(progress)
        }
      )

      if case .failure(let error) = result {
        throw Exception(
          name: "ConvertFailed",
          description: describeError(error)
        )
      }

      progressReporter.finish()
    }

    AsyncFunction("upload") {
      (filePath: String, url: String, config: UploadConfigRecord) async throws in
      let resolvedFilePath = resolveOutputPath(filePath)
      let progressReporter = ProgressReporter { progress in
        self.sendEvent("onProgress", ["progress": progress])
      }

      let result = await self.videoKit.upload(
        filePath: resolvedFilePath,
        url: url,
        config: config.toNative(),
        onProgress: { progress in
          progressReporter.updateNativeProgress(progress)
        }
      )

      if case .failure(let error) = result {
        throw Exception(
          name: "UploadFailed",
          description: describeError(error)
        )
      }

      progressReporter.finish()
    }

    AsyncFunction("convertAndUpload") {
      (
        inputUri: String,
        uploadUrl: String,
        outputPath: String?,
        conversionConfig: ConversionConfigRecord,
        uploadConfig: UploadConfigRecord
      ) async throws in
      let inputURL = try resolveURL(from: inputUri)
      let resolvedOutputPath = outputPath.map(resolveOutputPath(_:))
      let progressReporter = ProgressReporter { progress in
        self.sendEvent("onProgress", ["progress": progress])
      }

      let result = await self.videoKit.convertAndUpload(
        inputURL: inputURL,
        uploadURL: uploadUrl,
        outputPath: resolvedOutputPath,
        conversionConfig: conversionConfig.toNative(),
        uploadConfig: uploadConfig.toNative(),
        onProgress: { progress in
          progressReporter.updateNativeProgress(progress)
        }
      )

      if case .failure(let error) = result {
        throw Exception(
          name: "ConvertAndUploadFailed",
          description: describeError(error)
        )
      }

      progressReporter.finish()
    }
  }
}
